### PR TITLE
Modify model for ONNX compatibility

### PR DESCRIPTION
**Pull request by adrianlyjak**
This change implements fixes to allow exporting to ONNX from torch. I'm working on adding export scripts in a sibling repo at https://github.com/adrianlyjak/kokoro-onnx-export. I'm aware that there's ONNX models exported and shared on hugging face from taylorchu (github: https://github.com/taylorchu/kokoro-onnx, huggingface https://huggingface.co/onnx-community/Kokoro-82M-ONNX), but it seems like the export process hasn't been shared. I've been curious about learning more about the model, and learning about ONNX, so took a stab at this.

Not sure if you're interested in the changes here, or have any reservations about performance degradation or other such (not sure how the changes to the AdaIN1d might affect performance?). From what I can see, the vast majority of time and parameters is spent on the convolutions, so it doesn't seem like a significant change.

Summary of changes:

Simple fixes:
- removes usages of numpy conversions
- removes device `.to` movement, instead initializing tensors on the correct device from the start (Some of this is still necessary, for the rnn pack/unpack)

Moderate fixes:
- modifies/inlines normalization in AdaIN1d. Not entirely sure whether this is necessary for all code paths, but this fixes an issue with the layer when inputs were directly taken from an LSTM output
- modifies a permute call to remove a negative index, which was throwing off the onnx export 🤷 
- refactors the KModel interface, such that there's a good entrypoint for ONNX export to trace, that does not include the tokenization

Complicated fixes:
- Torch is moving towards using the dynamo backend for onnx exports. There's a bit of a catch 22 in torch onnx export functionality: dynamo based exports lack support for LSTM layers, which there are a number of, however it supports complex numbers, which the `TorchSTFT` uses. I was considering making a process to export parts of the model with the dynamo based export, and the other without, and gluing them together. However, the STFT was not too difficult to rewrite (I'm a little over my head here, this was mostly written by OpenAI o1), so I opted for that instead, and made this component optional to the onnx export only

### CODE REVIEW

**Comment by adrianlyjak** [comment]
@hexgrad 

> (e.g. a CustomSTFT implementation) should probably be forked off into separate classes such that Torch remains unaffected

Yeah, that's separated out, and is parameterized as an option that defaults to false. Is that sufficient?

> Has this been tested to perform a working ONNX export from the published Torch checkpoint?

Yup! It's working. I had to track down some issues in the STFT implementation (apparently tan2 has slightly differing behavior in onnx). The linked repository (while a work in progress) has CLI tools to export. The quality seems identical to my ear. I can't upload the wav file to github, but if you're interested, I can convert it to another file type and share here

**Comment by adrianlyjak** [comment]
Oh, for the AdaIN1d, do you have a preference between me doing performance testing? Or just going straight to making that a configurable option to only be used for the ONNX export

**Comment by adrianlyjak** [comment]
Oh, my mistake. There is some sort of a numerical difference between the InstanceNorm1d operation between pytorch and onnx, however its not a dealbreaker. It's _very_ slightly affecting the output duration, which was throwing off my comparisons. Seems better to revert that change, I'll instead just fix the comparisons

File: kokoro/model.py [hunk 1]
```diff
@@ -104,5 +103,37 @@ def forward(
...
+        assert len(input_ids)+2 <= self.context_length, (len(input_ids)+2, self.context_length)
+        input_ids = torch.LongTensor([[0, *input_ids, 0]]).to(self.device)
+        ref_s = ref_s.to(self.device)
+        audio, pred_dur = self.forward_with_tokens(input_ids, ref_s, speed)
+        audio = audio.squeeze().cpu()
+        pred_dur = pred_dur.cpu() if pred_dur is not None else None
+        logger.debug(f"pred_dur: {pred_dur}")
+        return self.Output(audio=audio, pred_dur=pred_dur) if return_output else audio
+
+class KModelForONNX(torch.nn.Module):
```
Comment (adrianlyjak) [inline kokoro/model.py:127]: I can take this out if it seems extra. Might be more appropriate with the export scripts. I had originally planned on including the export scripts in this change and target this repository, but there's a lot of churn and work there to do still

**Comment by adrianlyjak** [comment]
Thanks, sounds good to me!

**Comment by hexgrad** [comment]
LGTM, merging! Thanks for the PR
